### PR TITLE
Handle spurious EAGAIN when using fcntl() file locking

### PR DESCRIPTION
**Pull request by magnumripper**
We should never get EAGAIN with F_SETLKW, but a user saw it, see #3956.

### CODE REVIEW

**Comment by magnumripper** [comment]
@solardiz should I merge this as-is or could it have bad side-effects? We could make it a config option eg. `LockBugWorkaround = <bool>`.

**Comment by magnumripper** [comment]
Also, perhaps some sub-second sleeping (only when we see `EAGAIN`, before re-trying) would decrease the hammering on that poor cluster server?

**Comment by magnumripper** [comment]
Also, perhaps we should do the same workaround for `flock()`? User tried using flock instead and saw similar errors.

**Comment by solardiz** [comment]
I think we should:

1. Sleep for a little while using select(2) on `EAGAIN`, yet continue to retry immediately on `EINTR`.

2. Drop `flock` support entirely. Why do we still have it at all?

But I don't mind us merging this PR as-is first, then making the changes above separately.

The issue the user is having is probably a limit on the number of concurrently held locks.

**Comment by solardiz** [comment]
> The issue the user is having is probably a limit on the number of concurrently held locks.

Also, we might not want to fully mute the issue. Printing a warning (also explicitly suggesting that a limit on the number of concurrently held locks be raised?) the first time we hit the `EAGAIN` case could be useful.

**Comment by magnumripper** [comment]
> 2\. Drop `flock` support entirely. Why do we still have it at all?

I wouldn't know. Drop it in core and I'll follow! :-)

Anyway I'll do a more thorough patch instead of merging this:
- I'll move all our locks into a single shared helper function (in `logger.c`, no?) so we can do everything right in **one** place.
- Adding `log_event()` and print to stderr (both once per node). The screen output can suggest checking ulimits etc.
- I'm not sure how I'd use `select(2)` here. Would I put the affected fd in `errorfds`? I need to read up on that.

**Comment by magnumripper** [comment]
> * I'm not sure how I'd use `select(2)` here. Would I put the affected fd in `errorfds`? I need to read up on that.

Are you referring to something like this? https://stackoverflow.com/questions/3125645/why-use-select-instead-of-sleep

**Comment by magnumripper** [comment]
Done. For the sleep, I use `rand()` (which should do fine for this task, seeded with node number) as in
```c
			/* Sleep for a random time of max. 130 ms */
			t.tv_sec = 0; t.tv_usec = (rand() & 1023) << 7;
			select(0, NULL, NULL, NULL, &t);
			continue;
```
This is 0-130 ms, is that fine? I have no idea of what time frames we have here.

**Comment by magnumripper** [comment]
> This is 0-130 ms, is that fine? I have no idea of what time frames we have here. Perhaps once we hit this wall, a much longer wait would do more good and hopefully separate the nodes in time so they clash less at next event.

Hmm come to think of it, the timers wont separate anyway - next `timer_save` will clash just as much, right? Some ideas:
1. The **initial** `timer_save_value` in `sig_init()` could be ` += (NODE & 31)` (or 63) for 32x or 64x better separation without any other drawback than that the very first rec & log save with be up to 31 or 63 seconds later.
2. And/or we could add a new function in `signals.c` that merely does `timer_save_value++;` - and call that function whenever we get an `EAGAIN` here. Perhaps even when we get a `EINTR`?
3. And/or we could reset `timer_save_value` *after* files are saved instead of when setting `event_save = event_pending = 1`.
4. We could automagically increase the `LOG_BUFFER_SIZE` buffers according to total number of nodes.

@solardiz what do think?

**Comment by solardiz** [comment]
This is a lot of questions and a significant rework of the locking. I'll plan on commenting on this in the next few days.

File: src/logger.c [hunk 1]
```diff
@@ -113,6 +115,96 @@ static struct log_file pot = {NULL, NULL, NULL, 0, -1};
...
+ * params.
+ */
+int real_lock(int fd, int type, int mode,
+              const char *function, const char *file, int line)
+{
+#if OS_FLOCK || FCNTL_LOCKS
+
+#ifdef LOCK_DEBUG
+	fprintf(stderr, "%s(%u): Locking %s...\n",
+	        function, options.node_min, f->name);
```
Comment (magnumripper) [inline src/logger.c]: Fixme: We no longer have `f->name` so enabling LOCK_DEBUG wont work.

File: src/logger.c [hunk 2]
```diff
@@ -113,6 +115,96 @@ static struct log_file pot = {NULL, NULL, NULL, 0, -1};
...
+			continue;
+		} else if (errno != EINTR)
+			pexit(LOCK_FUNC "(%s%s)",
+			      type ? "exclusive lock" : "shared lock",
+			      mode ? "wait" : "nowait");
+	}
+
+#ifdef LOCK_DEBUG
+	fprintf(stderr, "%s(%u): Locked %s exclusively\n",
+	        __FUNCTION__, options.node_min, f->name);
```
Comment (magnumripper) [inline src/logger.c]: Same `f->name` problem

File: src/logger.c [hunk 3]
```diff
@@ -113,6 +115,96 @@ static struct log_file pot = {NULL, NULL, NULL, 0, -1};
...
+ * Shared helper for releasing a lock. Caller uses jtr_unlock(fd) which is
+ * actually a macro adding __FUNCTION__, __FILE__ and __LINE__ to params.
+ */
+void real_unlock(int fd, const char *function, const char *file, int line)
+{
+#if OS_FLOCK || FCNTL_LOCKS
+
+#ifdef LOCK_DEBUG
+	fprintf(stderr, "%s(%u): Unlocking %s\n",
+	        __FUNCTION__, options.node_min, f->name);
```
Comment (magnumripper) [inline src/logger.c]: Same `f->name` problem

File: src/logger.c (hunk 1 above)
Comment (magnumripper) [reply src/logger.c]: https://stackoverflow.com/a/13544447

Comment (magnumripper) [reply src/logger.c]: Not sure if there's a portable way of doing this at all

**Comment by magnumripper** [comment]
I'll rewrite this to have a single function and no `jtr_unlock()`

**Pull request closed without merging**